storage/balance_storage.go

```diff
@@ -673,3 +821,93 @@ func (b *BalanceStorage) SetBalanceImported(
 	log.Printf("%d Balances Updated\n", len(accountBalances))
 	return nil
 }
+
+// getHistoricalBalance returns the balance of an account
+// at a particular *types.BlockIdentifier.
+func (b *BalanceStorage) getHistoricalBalance(
+	ctx context.Context,
+	dbTx DatabaseTransaction,
+	account *types.AccountIdentifier,
+	currency *types.Currency,
+	block *types.BlockIdentifier,
+) (*types.Amount, *types.BlockIdentifier, error) {
+	var foundAmount *types.Amount
+	var foundBlock *types.BlockIdentifier
+	_, err := dbTx.Scan(
+		ctx,
+		GetHistoricalBalancePrefix(account, currency),
+		GetHistoricalBalanceKey(account, currency, block.Index),
+		func(k []byte, v []byte) error {
+			var deserialBal balanceEntry
+			// We should not reclaim memory during a scan!!
+			err := b.db.Encoder().Decode(historicalBalanceNamespace, v, &deserialBal, false)
+			if err != nil {
+				return fmt.Errorf(
+					"%w: unable to parse balance entry for %s",
+					err,
+					string(v),
+				)
+			}
+
+			// Ensure block hash matches in case of orphan
+			if deserialBal.Block.Index == block.Index && deserialBal.Block.Hash != block.Hash {
+				return fmt.Errorf(
+					"wanted block identifier %s but got %s",
+					types.PrintStruct(block),
+					types.PrintStruct(deserialBal.Block),
+				)
+			}
+
+			foundAmount = deserialBal.Amount
+			foundBlock = deserialBal.Block
+			return errAccountFound
```

Comment:
hmm why are we returning an error if the account is found? Is it limited by the signature of the Scan method?

Reply:
This is how we exit the query loop...otherwise we'd need to scan all entries unnecessarily